docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,225 @@
+
+This is the CRD that has a set of components to be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (only the clusters in the same namespace with the ClusterResourceSet resource).
+ClusterResourceSet is namespace-scoped, all resources and clusters needs to be in the same namespace as the ClusterResourceSet.
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: crs-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
+   - name: calico-addon
+     kind: Secret
+   - name: network-policy-addon
+     kind: ConfigMap
+status:
+ LastUpdated:  "2020-05-05T08:24:17Z"
+```
+
+Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Reconcile` mode that reapplies the resources on resource hash change and/or periodically.
+If ClusterResourceSet resources will be managed by an operator after they are applied by ClusterResourceSet controller, "ApplyOnce" mode must be used so that reconciliation on those resources can be delegated to the operator.
+
+Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.
+
+*Sample Secret Format*
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+type: clusterresourceset
+stringData:
+  value: |-
+    kind: Secret
+    apiVersion: v1
+    metadata:
+     name: mysql-secret
+```
+
+For preventing all secrets to be reached by all clusters in a namespace, only secrets with type "clusterresourceset" can be accessed by ClusterResourceSet controller.
+
+*Sample ConfigMap Format*
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: calico-addon
+data:
+  value: |-
+    kind: ConfigMap
+    apiVersion: v1
+    metadata:
+     name: calico-conf
+```
+
+The resources in ClusterResourceSet will be applied to matching clusters.
+There is many-to-many mapping between Clusters and ClusterResourceSets: Multiple ClusterResourceSets can match with a cluster; and multiple clusters can match with a single ClusterResourceSet.
+To keep information on which resources applied to which clusters, a new CRD is used, ClusterResourceSetStatus will be created in the management cluster. There will be one ClusterResourceSetStatus per workload cluster.
+
+Example:
+```yaml
+apiVersion: v1
+kind: ClusterResourceSetStatus
+metadata:
+  name: <cluster-name>
+  namespace: <cluster-namespace>
+spec:
+clusterresourcesets:
```

Comment:
I think this can be simplified by having a simple map of resources; this will also address cases when the same secret/configMap is part of more than one ResourceSet

Reply:
I think this may cause racing condition if two ClusterResourceSets have same resource and they need to change this from different controller threads. Wdyt?

Reply:
I think that i two resource set have the same resource, the resource should be applied only once. In fact, this structure helps to implement this behavior